### PR TITLE
Fix statevector being sent to Narnia by CRy(pi)

### PR DESCRIPTION
**Pull request by ausbin**
I believe there is a bug in the sparse simulator when you run the circuit $(X \otimes I)CRy(\pi)(X Ry(\pi/2) \otimes I)\ket{00}$. The expected output of the simulator is $\frac{1}{\sqrt{2}}(\ket{01} + \ket{10})$, but instead the state becomes the malformed state $\frac{1}{\sqrt{2}}\ket{10}$.

Here is the QIR I used to trigger this (note this is LLVM 19 IR, so it has opaque pointers... sorry if you are on an older LLVM version):

<details>
<summary>LLVM 19 IR Example</summary>

```llvm
; ModuleID = 'LLVMDialectModule'
source_filename = "LLVMDialectModule"

@tag_ret = internal constant [4 x i8] c"ret\00"
@tag_discarded = internal constant [10 x i8] c"discarded\00"

define i64 @kernel_0() #0 {
  call void @__quantum__rt__initialize(ptr null)
  call void @__quantum__qis__ry__body(double 0x3FF921FB54442D18, ptr null)
  call void @__quantum__qis__x__body(ptr null)
  %1 = call ptr @__quantum__rt__array_create_1d(i32 ptrtoint (ptr getelementptr (ptr, ptr null, i32 1) to i32), i64 1)
  %2 = call ptr @__quantum__rt__array_get_element_ptr_1d(ptr %1, i64 0)
  store ptr null, ptr %2, align 8
  %3 = call ptr @__quantum__rt__tuple_create(i64 ptrtoint (ptr getelementptr ({ double, ptr }, ptr null, i32 1) to i64))
  %4 = getelementptr { double, ptr }, ptr %3, i32 0, i32 0
  store double 0x400921FB54442D18, ptr %4, align 8
  %5 = getelementptr { double, ptr }, ptr %3, i32 0, i32 1
  store ptr inttoptr (i64 1 to ptr), ptr %5, align 8
  call void @__quantum__qis__ry__ctl(ptr %1, ptr %3)
  call void @__quantum__rt__array_update_reference_count(ptr %1, i32 -1)
  call void @__quantum__rt__tuple_update_reference_count(ptr %3, i32 -1)
  call void @__quantum__qis__x__body(ptr null)
  call void @__quantum__qis__mz__body(ptr null, ptr null)
  call void @__quantum__qis__mz__body(ptr inttoptr (i64 1 to ptr), ptr inttoptr (i64 1 to ptr))
  call void @__quantum__rt__tuple_record_output(i64 2, ptr @tag_ret)
  call void @__quantum__rt__result_record_output(ptr null, ptr null)
  call void @__quantum__rt__result_record_output(ptr inttoptr (i64 1 to ptr), ptr null)
  call void @__quantum__rt__tuple_record_output(i64 0, ptr @tag_discarded)
  ret i64 0
}

declare void @__quantum__qis__x__body(ptr)

declare void @__quantum__qis__ry__body(double, ptr)

declare void @__quantum__qis__ry__ctl(ptr, ptr)

declare ptr @__quantum__rt__array_create_1d(i32, i64)

declare ptr @__quantum__rt__array_get_element_ptr_1d(ptr, i64)

declare void @__quantum__rt__array_update_reference_count(ptr, i32)

declare ptr @__quantum__rt__tuple_create(i64)

declare void @__quantum__rt__tuple_update_reference_count(ptr, i32)

declare void @__quantum__rt__initialize(ptr)

declare void @__quantum__qis__mz__body(ptr, ptr)

declare void @__quantum__rt__tuple_record_output(i64, ptr)

declare void @__quantum__rt__result_record_output(ptr, ptr)

attributes #0 = { "entry_point" "output_labeling_schema"="qwerty_v1" "qir_profiles"="base_profile" "required_num_qubits"="2" "required_num_results"="2" }

!llvm.module.flags = !{!0}

!0 = !{i32 2, !"Debug Info Version", i32 3}
```

</details>

I also added a unit test for this. I hope the fix is as clear as I think it is and I'm not misunderstanding the code. (My reading is that amplitudes for basis states not matching the control bits are being left out entirely instead of being left alone as intended.)

### CODE REVIEW

File: sparsesim/src/lib.rs [hunk 1]
```diff
@@ -1074,6 +1074,8 @@ impl QuantumSim {
                         .fold(SparseState::default(), |mut accum, (index, value)| {
                             if ctls.iter().all(|c| index.bit(*c)) {
                                 accum.insert(index, value * factor);
+                            } else {
```
Comment (swernli) [inline sparsesim/src/lib.rs:1077]: Good catch, thanks for the fix! It looks like I made the same mistake on line 1091 below as well. Could you apply the same fix there too?

**Comment by swernli** [comment]
@microsoft-github-policy-service rerun

File: sparsesim/src/lib.rs (hunk 1 above)
Comment (ausbin) [reply sparsesim/src/lib.rs:1077]: Yes sir, just attempted that. What do you think?

File: sparsesim/src/lib.rs [hunk 2]
```diff
@@ -1074,20 +1074,25 @@ impl QuantumSim {
...
+                                accum.insert(index, value);
                             }
                             accum
                         });
             }
         } else if m01.is_nearly_zero() {
             // This is just identity, so we can effectively no-op, and just add a phase of -1 as indicated by m00.
-            if m00 == -Complex64::one() {
+            // Here, m00 + 1 == 0 is used to check if m00 == -1.
+            if (m00 + Complex64::one()).is_nearly_zero() {
```
Comment (ausbin) [inline sparsesim/src/lib.rs:1086]: Did this in case it's -0.99999999999999 or something

File: sparsesim/src/lib.rs (hunk 1 above)
Comment (swernli) [reply sparsesim/src/lib.rs:1077]: Looks great! Not sure why the pipeline is mad (looks like an intermittent issue with the license task, the tests are passing). I'll try to get this merged.

**Review by swernli — APPROVED** [review]

**Comment by swernli** [comment]
Looks like the pipeline issue is due to the fact that the repo still uses the deprecated macos-12 runners. I'll fix that in a separate PR and merge this with admin privileges.

**Comment by swernli** [comment]
@ausbin would it help to get a release out soon with this fix or do you have what you need to be unblocked on your end? If it's helpful we could do one this week, but otherwise it might be easier to wait until after the Thanksgiving holiday.

**Comment by ausbin** [comment]
Thank you for asking, @swernli, that's kind of you.

I do use qir-runner heavily, but I package it up in a harebrained way that isn't dependent on there being a qir-runner release or not.

Enjoy your Thanksgiving!

**Pull request merged**